src/GNDStk/gnds-utility.hpp

```diff
@@ -0,0 +1,231 @@
+
+// -----------------------------------------------------------------------------
+// Macros, variables, enums
+// -----------------------------------------------------------------------------
+
+// printval
+// printvar
+// Are the same, so we don't need to remember the exact terminology
+#define printval(var) std::cout << #var ": [" << var << "]" << std::endl
+#define printvar(var) printval(var)
+
+// indent
+// Number of spaces of indentation you want, in the output of certain types
+inline int indent = 3;
+
+// format
+enum class format {
+   null, // default, automagick, etc.
+   tree, // our own simple text format
+   xml,
+   json,
+   hdf5
+};
+
+
+
+// -----------------------------------------------------------------------------
+// Forward declarations: general
+// -----------------------------------------------------------------------------
+
+// fixme Some reorganization elsewhere may make some of these unnecessary
+
+// Node, node
+template<
+   template<class...> class,
+   template<class...> class
+>
+class Node;
+using node = Node<std::vector,std::vector>;
+
+// Tree, tree
+template<
+   template<class...> class = std::vector,
+   template<class...> class = std::vector
+>
+class Tree;
```

Comment:
What's the difference between `Tree`, `node`, and `tree`? They all seem the same to me.

Reply:
Tree (capital T) is my templated version. tree (lower-case t) is Tree<default arguments>. These are what we talked about in your office; I know you'd like renaming to Tree_t and Tree, respectively. Same difference with Node and node. The distinction between trees and nodes, as I spoke of when we met, is there for what I believe is a good reason. They could be forced into the same type, but I think that keeping them separate is far better design.

Reply:
Or course that second sentence is supposed to say ```Tree<>```, not Tree. Github smart-formatting strikes [out] again. Sort of like auto-incorrect on my phone.

Reply:
I had a conversation recently with a computer scientist friend, and we got to talking about trees and the various ways one might represent them. Touching on the "node versus tree" concept, he made what I think was an excellent comparison: that of a mathematical problem versus an engineering problem.

From a largely mathematical concept, and drawing a picture, one could say the a tree's top-level node looks the same, structurally, as every other node. They're all about the same.  From a purist standpoint, just make one class.

From more of an engineering standpoint, we should ask whether making them the same would allow users to more easily make mistakes - say, doing something with a (non-top-level) node that really only makes sense with an entire GNDS tree. Also - and I think I said as much in a conversation at one time - I anticipate that some worthwhile (efficiency- or feature-wise) extra data could be put in the general tree structure, while such data would just be clutter, and not useful, in individual nodes. (Or the nodes might deserve different extra information.) It's true that there is, at the moment, some code replication or near-replication in nodes versus trees, and that's something I plan to look at more closely sometime, especially after tests are more complete, and users have had a chance to begin using the tool and providing feedback. Deriving tree from node might be an option; or deriving both from something else. At the moment I think the functionality works well, and would suggest that we not churn it up prematurely.